File: WdtcCore/src/main/java/org/wdt/wdtc/download/fabric/FabricVersionList.java
package org.wdt.wdtc.download.fabric;


import org.wdt.wdtc.download.infterface.VersionList;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.utils.URLUtils;
import org.wdt.wdtc.utils.gson.JSONArray;
import org.wdt.wdtc.utils.gson.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FabricVersionList implements VersionList {

    @Override
    public List<String> getVersionList() throws IOException {
        List<String> FabricVersionList = new ArrayList<>();
        JSONArray list = JSONArray.parseJSONArray(URLUtils.getURLToString(Launcher.getDownloadSource().getFabricMetaUrl() + "v2/versions/loader"));
        for (int i = 0; i < list.size(); i++) {
            JSONObject fabricObject = list.getJSONObject(i);
            FabricVersionList.add(fabricObject.getString("version"));
        }
        return FabricVersionList;
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/fabric/FabricVersionList.kt
package org.wdt.wdtc.core.impl.download.fabric

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.openapi.download.interfaces.VersionListInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asVersionsObjectSequence
import org.wdt.wdtc.core.openapi.manger.currentDownloadSource
import org.wdt.wdtc.core.openapi.utils.ioAsync
import org.wdt.wdtc.core.openapi.utils.toURL
import java.util.*

class FabricVersionList : VersionListInterface {
	private val versionListUrl = "${currentDownloadSource.fabricMetaUrl}v2/versions/loader"
	
	private val list = ioAsync { versionListUrl.toURL().toStrings().parseJsonArray() }
	
	override suspend fun getVersionList(): VersionsObjectSequence {
		return list.await().asSequence<FabricVersionsJsonObjectImpl>().asVersionsObjectSequence()
	}
	
	class FabricVersionsJsonObjectImpl(
		@field:SerializedName("version")
		override var versionNumber: String,
		
		@field:SerializedName("build")
		val buildNumber: Int = 0
	
	) : VersionsJsonObjectInterface {
		
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (other == null || javaClass != other.javaClass) return false
			val that = other as FabricVersionsJsonObjectImpl
			return buildNumber == that.buildNumber && versionNumber == that.versionNumber
		}
		
		override fun hashCode(): Int {
			return Objects.hash(versionNumber, buildNumber)
		}
		
		
	}
	
}
